Bluetooth: controller: nRF5: Back-to-Back Radio Tx interface

cvinayak opened the pull request:
Add Radio interface to perform back-to-back transmit of PDU
with a configurable inter frame spacing.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: nRF5: Back-to-Back Radio Tx interface

Add Radio interface to perform back-to-back transmit of PDU
with a configurable inter frame spacing.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

andrzej-kaczmarek reviewed (approved):
lgtm

tested with https://github.com/zephyrproject-rtos/zephyr/pull/31875

kruithofa commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -565,7 +575,7 @@ void radio_switch_complete_and_rx(uint8_t phy_rx)
 	NRF_RADIO->SHORTS = RADIO_SHORTS_READY_START_Msk |
 			    RADIO_SHORTS_END_DISABLE_Msk;
-	sw_switch(0, 0, 0, phy_rx, 0);
+	sw_switch(1, 0, 0, 0, phy_rx, 0);
```
To improve readability: can we use define's instead of hardcoded 1/0

kruithofa commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -579,7 +589,21 @@ void radio_switch_complete_and_tx(uint8_t phy_rx, uint8_t flags_rx, uint8_t phy_
 	NRF_RADIO->SHORTS = RADIO_SHORTS_READY_START_Msk |
 			    RADIO_SHORTS_END_DISABLE_Msk;
-	sw_switch(1, phy_rx, flags_rx, phy_tx, flags_tx);
+	sw_switch(0, 1, phy_rx, flags_rx, phy_tx, flags_tx);
```
To improve readability: can we use define's instead of hardcoded 1/0

kruithofa commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -579,7 +589,21 @@ void radio_switch_complete_and_tx(uint8_t phy_rx, uint8_t flags_rx, uint8_t phy_
+#else /* !CONFIG_BT_CTLR_TIFS_HW */
+	NRF_RADIO->SHORTS = RADIO_SHORTS_READY_START_Msk |
+			    RADIO_SHORTS_END_DISABLE_Msk;
+	sw_switch(1, 1, phy_curr, flags_curr, phy_next, flags_next);
```
To improve readability: can we use define's instead of hardcoded 1/0

kruithofa commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -446,24 +446,34 @@ void *radio_pkt_decrypt_get(void)
-		/* Calculate delay with respect to current (RX) and next
-		 * (TX) PHY. If RX PHY is LE Coded, assume S8 coding scheme.
+		/* Calculate delay with respect to current and next PHY.
+		 * If RX PHY is LE Coded, assume S8 coding scheme.
```
Is the comment still correct? If (dir_curr !=0) then no such assumption is made, but flags_curr is used, right?

cvinayak committed:
Bluetooth: controller: Enclose macro parameters in paranthesis

Enclose macro parameters in paranthesis to avoid ambiguous
macro expansions.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Minor indentation and HAL code change

Minor indentation change and replaced if-then-else-if clause
with toggle implementation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: nRF53 sw_switch always required Radio End

nRF53 implementation of sw_switch always requires Radio End
event, hence optimize out redundant code due to explicit use
of radio_tmr_end_capture.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -447,28 +447,55 @@ void *radio_pkt_decrypt_get(void)
 
 #if !defined(CONFIG_BT_CTLR_TIFS_HW)
+
+#define SW_SWITCH_PREV_RX             0
```
`SW_SWITCH_DIR_*` used for both TX and RX would be less confusing

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -447,28 +447,55 @@ void *radio_pkt_decrypt_get(void)
+#define SW_SWITCH_NEXT_RX             0
+#define SW_SWITCH_PREV_TX             1
+#define SW_SWITCH_NEXT_TX             1
+#define SW_SWITCH_PREV_PHY_1M         0
```
better use `PHY_LEGACY` or even better `PHY_1M` instead since this is actual value passed from LLL

cvinayak replied:
Same here, macros are used as readable string (instead of using defines) which would convey the parameters purpose. Otherwise, the call at line 607 will look like:
```
sw_switch(SW_SWITCH_DIR_TX, SW_SWITCH_DIR_RX,
		  PHY_1M, SW_SWITCH_PREV_FLAGS_DONTCARE,
		  phy_rx, SW_SWITCH_NEXT_FLAGS_DONTCARE);
```

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -447,28 +447,55 @@ void *radio_pkt_decrypt_get(void)
+#define SW_SWITCH_NEXT_TX             1
+#define SW_SWITCH_PREV_PHY_1M         0
+#define SW_SWITCH_PREV_FLAGS_DONTCARE 0
+#define SW_SWITCH_NEXT_FLAGS_DONTCARE 0
```
could be defined in `pdu.h` and eventually also used in LLL since flags are magic values there as well

cvinayak replied:
Again the macros are replacement to hardcoded values passed to `sw_switch` function which is vendor specific interface, not sure if other vendors will use `0`  as the dont care values.

Change other magic values is outside the scope of this PR.

andrzej-kaczmarek reviewed (commented):
minor comments, verified again with https://github.com/zephyrproject-rtos/zephyr/pull/31875 and works fine

cvinayak reviewed (commented):
This PR depends on #31336 as that PR has resolved the conflicts of having a separate _DPPI.h header file introduced in #30732.

I have resolved conflicts in this PR over #31336 in my another branch, hence it will be less work if #31336 is merged first.

carlescufi commented:
@andrzej-kaczmarek can you please take another look?

cvinayak commented:
This needs a rebase and also would like https://github.com/zephyrproject-rtos/zephyr/pull/31336 to go first to avoid more working resolving conflicts
